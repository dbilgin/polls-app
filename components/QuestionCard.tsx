import React from 'react';
import {StyleSheet, View, Text, Image, ImageSourcePropType} from 'react-native';
import {Question} from '../Models/Question';

interface CardProps {
  q: Question;
  image: ImageSourcePropType;
}

const QuestionCard: React.FC<CardProps> = ({q, image}) => {
  function sum(obj: number[]) {
    let finalSum = 0;
    for (const el of obj) {
      finalSum += el;
    }
    return finalSum;
  }

  return (
    <View style={styles.cardContainer}>
      <Image
        style={{flex: 0.5, width: '100%'}}
        resizeMode={'contain'}
        source={image}
      />
      <View style={{padding: 24, flex: 1}}>
        <Text style={styles.title}>{q.question}</Text>
        <Text style={styles.subtitle}>{q.url}</Text>
        <Text style={styles.dateText}>
          Published at: {new Date(q.published_at).toLocaleDateString()}
        </Text>
        <Text style={styles.bottomText}>{`This question has ${
          q.choices.length
        } choices and a total of ${sum(
          q.choices.map((x) => x.votes),
        )} votes. Press here now to vote!`}</Text>
      </View>
    </View>
  );
};

export default React.memo(QuestionCard);

const styles = StyleSheet.create({
  bottomText: {
    fontSize: 24,
    color: 'black',
    marginTop: 24,
  },
  dateText: {
    fontSize: 16,
    color: '#505b7d',
    marginTop: 12,
  },
  subtitle: {
    fontSize: 12,
    color: '#505b7d',
    marginTop: 12,
  },
  title: {
    fontSize: 24,
    color: '#505b7d',
  },
  cardContainer: {
    height: 500,
    width: '80%',
    margin: 20,
    borderWidth: 0,
    borderRadius: 18,
    justifyContent: 'center',
    alignSelf: 'center',
    backgroundColor: 'white',

    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 4,
    },
    shadowOpacity: 0.3,
    shadowRadius: 4.65,

    elevation: 8,
  },
});
