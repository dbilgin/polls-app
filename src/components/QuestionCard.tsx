import React from 'react';
import {
  StyleSheet,
  View,
  Text,
  Image,
  ImageSourcePropType,
  TouchableOpacity,
  SafeAreaView,
  Alert,
} from 'react-native';
import {Question} from '../../Models/Question';
import {sum} from '../utils';
import {RootStackParamList} from '../../App';
import {StackNavigationProp} from '@react-navigation/stack';

interface CardProps {
  q: Question;
  image: ImageSourcePropType;
  navigation: StackNavigationProp<RootStackParamList, 'QuestionListScreen'>;
}

const QuestionCard: React.FC<CardProps> = ({q, image, navigation}) => {
  const navigateToDetail = () => {
    if (q.id) {
      navigation.navigate('QuestionDetailScreen', {questionId: q.id});
    } else {
      Alert.alert('An error has occurred.');
    }
  };

  return (
    <SafeAreaView style={{flex: 1}}>
      <TouchableOpacity onPress={navigateToDetail} style={styles.cardContainer}>
        <Image
          style={{flex: 0.5, width: '100%'}}
          resizeMode={'contain'}
          source={image}
        />
        <View style={{padding: 24, flex: 1}}>
          <Text style={styles.title}>{q.question}</Text>
          <Text style={styles.subtitle}>{q.url}</Text>
          <Text style={styles.dateText}>
            Published at: {new Date(q.published_at).toLocaleDateString()}
          </Text>
          <Text style={styles.bottomText}>{`This question has ${
            q.choices.length
          } choices and a total of ${sum(
            q.choices.map((x) => x.votes),
          )} votes. Press here now to vote!`}</Text>
        </View>
      </TouchableOpacity>
    </SafeAreaView>
  );
};

export default QuestionCard;

const styles = StyleSheet.create({
  bottomText: {
    fontSize: 24,
    color: 'black',
    marginTop: 24,
  },
  dateText: {
    fontSize: 16,
    color: '#505b7d',
    marginTop: 12,
  },
  subtitle: {
    fontSize: 12,
    color: '#505b7d',
    marginTop: 12,
  },
  title: {
    fontSize: 24,
    color: '#505b7d',
  },
  cardContainer: {
    height: 500,
    width: '80%',
    margin: 20,
    borderWidth: 0,
    borderRadius: 18,
    justifyContent: 'center',
    alignSelf: 'center',
    backgroundColor: 'white',

    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 4,
    },
    shadowOpacity: 0.3,
    shadowRadius: 4.65,

    elevation: 8,
  },
});
