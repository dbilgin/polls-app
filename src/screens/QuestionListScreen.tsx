import React, {useEffect} from 'react';
import {useDispatch, useSelector} from 'react-redux';
import {
  StyleSheet,
  View,
  Text,
  ActivityIndicator,
  TouchableOpacity,
  SafeAreaView,
} from 'react-native';
import {fetchQuestion} from '../store/actions';
import {RootState} from '../store/reducers';
import {Question} from '../../Models/Question';
import QuestionCard from '../components/QuestionCard';
import {StackNavigationProp} from '@react-navigation/stack';
import {RootStackParamList} from '../../App';
import {FAB} from 'react-native-paper';
import {debounce} from 'lodash';
import Swiper from 'react-native-swiper';

const images = [
  require('../images/1.png'),
  require('../images/2.png'),
  require('../images/3.png'),
];

type QuestionListScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'QuestionListScreen'
>;

type Props = {
  navigation: QuestionListScreenNavigationProp;
};

const QuestionListScreen: React.FC<Props> = ({navigation}) => {
  const dispatch = useDispatch();
  const qs = useSelector((state: RootState) => state.questions);

  useEffect(() => {
    dispatch(fetchQuestion(1));
    dispatch(fetchQuestion(2));
  }, [dispatch]);

  const renderFooter = () => {
    return (
      <SafeAreaView>
        <View style={{height: 50}}>
          {qs.isFetching ? (
            <ActivityIndicator color="black" />
          ) : !!qs.error && qs.error?.response?.status !== 404 ? (
            <TouchableOpacity
              style={styles.errorButton}
              onPress={() => debounceRequestNewPage()}>
              <Text style={styles.loadText}>
                An error has occurred, press here to retry
              </Text>
            </TouchableOpacity>
          ) : null}
        </View>
      </SafeAreaView>
    );
  };

  const renderItem = ({item, index}: {item: Question; index: number}) => {
    let newIndex = index;
    while (newIndex > 2) {
      newIndex = newIndex - 3;
    }
    const randomImage = images[newIndex];
    return (
      <QuestionCard
        key={item.url + 'card'}
        q={item}
        image={randomImage}
        navigation={navigation}
      />
    );
  };

  const endReached = () => {
    if (qs.error?.response?.status !== 404) {
      debounceRequestNewPage();
    }
  };

  const debounceRequestNewPage = debounce(() => {
    const maxId = Math.max.apply(
      Math,
      qs.questions.map((o) => {
        return o.id ?? 0;
      }),
    );

    dispatch(fetchQuestion(maxId + 1));
  }, 100);

  return (
    <View style={styles.container}>
      {qs.questions.length > 1 && (
        <Swiper
          loop={false}
          showsPagination={false}
          showsButtons={true}
          onIndexChanged={(index) => {
            if (qs.questions.length - (index + 1) < 1) {
              endReached();
            }
          }}>
          {qs.questions.map((question: Question, index: number) =>
            renderItem({item: question, index: index}),
          )}
        </Swiper>
      )}

      <FAB
        style={styles.fab}
        icon="plus"
        onPress={() => navigation.navigate('QuestionAddScreen')}
      />
      {renderFooter()}
    </View>
  );
};

export default QuestionListScreen;

const styles = StyleSheet.create({
  fab: {
    position: 'absolute',
    margin: 30,
    right: 0,
    bottom: 0,
    backgroundColor: '#147efb',
  },
  container: {
    backgroundColor: 'white',
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  separator: {
    height: 0.5,
    backgroundColor: 'rgba(0,0,0,0.4)',
  },
  errorButton: {
    height: 50,
    backgroundColor: '#137177',
    justifyContent: 'center',
    alignItems: 'center',
  },
  loadText: {
    color: 'white',
  },
});
