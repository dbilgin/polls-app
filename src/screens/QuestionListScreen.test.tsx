import React from 'react';
import {shallow} from 'enzyme';
import {TouchableOpacity, ActivityIndicator, SafeAreaView} from 'react-native';
import QuestionListScreen from './QuestionListScreen';
import QuestionCard from '../components/QuestionCard';
import Swiper from 'react-native-swiper';

const createTestProps = () => ({
  route: {
    params: {
      questionId: 1,
    },
  },
  navigation: {
    navigate: jest.fn(),
    goBack: jest.fn(),
  },
});

jest.mock('../PollsService', () => ({
  voteQuestion: jest.fn(() => Promise.resolve({status: 201})),
}));

jest.mock('react-redux', () => ({
  useSelector: () => {
    return {
      questions: [
        {
          id: 1,
          question: 'question',
          choices: [{choice: 'choice 1', url: 'url/1', votes: 0}],
          published_at: new Date('2020-06-07T18:21:01.371Z'),
          url: '/url/1',
        },
        {
          id: 2,
          question: 'question',
          choices: [{choice: 'choice 1', url: 'url/1', votes: 0}],
          published_at: new Date('2020-06-07T18:21:01.371Z'),
          url: '/url/2',
        },
      ],
      isFetching: false,
      error: undefined,
    };
  },
  useDispatch: () => jest.fn(),
}));

describe('Rendering', () => {
  let props: any;
  beforeEach(() => {
    props = createTestProps();
  });

  it('should match to snapshot', () => {
    const component = shallow(<QuestionListScreen {...props} />);
    expect(component).toMatchSnapshot();
  });
});

describe('UI Placements', () => {
  let props: any;
  beforeEach(() => {
    props = createTestProps();
  });

  it('Data available - correct count of components', async () => {
    const component = shallow(<QuestionListScreen {...props} />);
    expect(component.find(TouchableOpacity).length).toBe(1);
    expect(component.find(ActivityIndicator).length).toBe(0);
    expect(component.find(QuestionCard).length).toBe(2);
    expect(component.find(Swiper).length).toBe(1);
    expect(component.find(SafeAreaView).length).toBe(1);
  });
});

describe('Interactions', () => {
  let props: any;
  beforeEach(() => {
    props = createTestProps();
  });

  it('Navigation', async () => {
    const component = shallow(<QuestionListScreen {...props} />);

    // @ts-ignore
    await component.find(TouchableOpacity).first().props().onPress();
    expect(props.navigation.navigate).toHaveBeenCalled();
  });
});
