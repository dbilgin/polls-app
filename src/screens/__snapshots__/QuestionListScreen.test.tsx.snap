// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering should match to snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <_default
    automaticallyAdjustContentInsets={false}
    autoplay={false}
    autoplayDirection={true}
    autoplayTimeout={2.5}
    bounces={false}
    disableNextButton={false}
    disablePrevButton={false}
    horizontal={true}
    index={0}
    loadMinimal={false}
    loadMinimalSize={1}
    loop={false}
    onIndexChanged={[Function]}
    pagingEnabled={true}
    removeClippedSubviews={true}
    scrollsToTop={false}
    showsButtons={true}
    showsHorizontalScrollIndicator={false}
    showsPagination={false}
    showsVerticalScrollIndicator={false}
  >
    <QuestionCard
      image={
        Object {
          "testUri": "../../../src/images/1.png",
        }
      }
      key="/url/1card"
      navigation={
        Object {
          "goBack": [MockFunction],
          "navigate": [MockFunction],
        }
      }
      q={
        Object {
          "choices": Array [
            Object {
              "choice": "choice 1",
              "url": "url/1",
              "votes": 0,
            },
          ],
          "id": 1,
          "published_at": 2020-06-07T18:21:01.371Z,
          "question": "question",
          "url": "/url/1",
        }
      }
    />
    <QuestionCard
      image={
        Object {
          "testUri": "../../../src/images/2.png",
        }
      }
      key="/url/2card"
      navigation={
        Object {
          "goBack": [MockFunction],
          "navigate": [MockFunction],
        }
      }
      q={
        Object {
          "choices": Array [
            Object {
              "choice": "choice 1",
              "url": "url/1",
              "votes": 0,
            },
          ],
          "id": 2,
          "published_at": 2020-06-07T18:21:01.371Z,
          "question": "question",
          "url": "/url/2",
        }
      }
    />
  </_default>
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[Function]}
    style={
      Object {
        "backgroundColor": "#147efb",
        "borderRadius": 50,
        "borderWidth": 0,
        "bottom": 0,
        "margin": 30,
        "padding": 15,
        "position": "absolute",
        "right": 0,
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/images/plus.png",
        }
      }
      style={
        Object {
          "height": 40,
          "tintColor": "white",
          "width": 40,
        }
      }
    />
  </TouchableOpacity>
  <SafeAreaView>
    <View
      style={
        Object {
          "height": 50,
        }
      }
    />
  </SafeAreaView>
</View>
`;
